### ISSUE TITLE
IR Interpreter optimization ideas

### ISSUE THREAD

**Issue by hrydgard** [hrydgard/ppsspp#19143]
Mainly, this will be about reducing the amount of instructions interpreted (to reduce interpreter overhead) and to make simpler versions of instructions that are faster to run.

* Merge Downcount into Exit when possible (yes even into the conditional ones). When downcount fits into a byte, we have space for that in all Exit instructions I think...
* Multi-load/multi-store instructions (to replace long sequences of say Load32 s0, sp, 0x14, Load32 s1, sp, 0x18, Load32 s0, sp, 0x1C, etc)
* Super-specialized instructions, for example:
  - AddConst sp, sp, 0x30 is very common, it might be very slightly beneficial to add AddConstSame sp, 0x30
  - sw zero, sp, 0x10 is quite common

There's a bit of tension between keeping the IR easy to run passes on, and making it fast to interpret.

Maybe we should translate further to an even more specialized interpreter IR, or, only ever use the super-specialized instructions in the very final optimization pass so only the interpreter needs to care about them. In that case they should be marked clearly.

**Comment by fp64** [comment]
> Write the IRInterpreter in assembler (hopefully not needed, but would love to get rid of the range check that the switch emits, at least on x86)

Wouldn't `__builtin_unreachable()`/`__assume(0)` work for getting rid of a range check? Something like https://godbolt.org/z/Msfsd63c6 (if that's what you mean)?

**Comment by hrydgard** [comment]
@fp64 hm yes, that's an interesting idea! I'm not too familiar with these tricks (since I thought they didn't work in VS) but sure seems like they can do the trick!

**Comment by hrydgard** [comment]
@fp64 Implemented here: https://github.com/hrydgard/ppsspp/pull/20875

**Comment by fp64** [comment]
`Vec4Blend` (unlike `Vec4Shuffle`) seems fairly straighforward (on x86 at least), e.g. https://godbolt.org/z/TxKnadoP4, so maybe less need to special-case?
Note that GCC produces much messier code than clang here, and neither would probably auto-vectorize actual `IROp::Vec4Blend`, since the current code writes directly to `mips->f[dest + i]` not to temprary, so compiler cannot assume "no partial overlap on src/dst" (which assumption, I hope, is actually aways true) and has to go scalar.

**Comment by hrydgard** [comment]
Hm, good point about overlap. And we should probably experiment with that vector annotation.. that output looks great.

Though no gcc-only platforms that I know of actually need the IR interpreter - iOS uses clang, and is ARM64 only...

and clang does pretty well with the initial code, with those vec4 annotations: https://godbolt.org/z/84a7Kc64M

without the annotation, on arm64, temps do indeed help code quality, not sure about speed: https://godbolt.org/z/a1K5bc8hY

**Comment by fp64** [comment]
For general shuffle, `switch` doesn't seem to be worth it (vs. reference): https://godbolt.org/z/183xs6anW.
`pshufb` may be, if we can actually inline it (GCC/clang wouldn't want to inline SSSE3 in the middle of SSE2 function). Without (and checking CPU support each time, even if check is just loading variable + well-predicted branch), it may be actually a slowdown.

**Comment by fp64** [comment]
For funsies, also tried self-modifying code: https://godbolt.org/z/va4nrMnez. Obviously impractical, was just curious how much slower it is. Turn out, about x50..x100.

**Comment by hrydgard** [comment]
Hehe, not unexpected.

Anyway, I don't want to require more than SSE2 on x86, though at this point it would probably be fine to require SSSE3 for the 64-bit build, at least...  There are only a rare few early x64 processors that don't support it.

But also, the IR interpreter is really only used on ARM64. Maybe looking into the VTBL instructions might be interesting..

**Comment by fp64** [comment]
Unless I'm missing something, `Vec4Pack31To8` seems straightforward: https://godbolt.org/z/ch18716bE (and `Vec4Pack32To8` even simpler, no `pand`).

**Comment by hrydgard** [comment]
Yeah. Although, I'm considering reworking those IR instructions a bit - it's always preceded by a ClampToZero so might as well merge those two, there are some similar cases where the original idea was to make these instructions easy for the next JIT (IR -> native), but in reality the IR has mostly become used for interpretation which has some different demands.

Also these weird instructions are only heavily used by a few games, but might still be worth optimizing.
